programs/mango-v4/src/instructions/serum3_place_order.rs

```diff
@@ -258,6 +258,48 @@ pub fn serum3_place_order(
         None
     };
 
+    max_base_qty = if serum_market.is_reduce_only() {
+        require!(
+            load_open_orders_ref(&ctx.accounts.open_orders.as_ref())?.free_slot_bits == u128::MAX,
+            MangoError::SomeError
+        );
+
+        let bank = ctx.accounts.payer_bank.load()?;
+        let token_pos_native = account
+            .token_position(bank.token_index)?
+            .native(&bank)
+            .to_num::<i64>();
+        let base_lot_size = load_market_state(
+            &ctx.accounts.serum_market_external,
+            &ctx.accounts.serum_program.key(),
+        )?
+        .coin_lot_size;
+        let max_base_qty_native = max_base_qty * base_lot_size;
+        require!(
+            max_base_qty_native >= token_pos_native.unsigned_abs(),
+            MangoError::SomeError
+        );
+
+        match side {
+            Serum3Side::Bid => {
+                if token_pos_native > 0 {
+                    0
+                } else {
+                    token_pos_native.min(0).unsigned_abs().min(max_base_qty)
+                }
+            }
+            Serum3Side::Ask => {
+                if token_pos_native < 0 {
+                    0
+                } else {
+                    (token_pos_native.max(0) as u64).min(max_base_qty)
+                }
+            }
+        }
+    } else {
+        max_base_qty
+    };
```

Comment:
Instead of this logic, I'd make it dependent on bank.is_reduce_only. Every place where we check the deposit_to_vault_ratio likely also needs to care about the bank's reduce-only state.

Reply:
Good point, I can actually drop the serum3_market.is_reduce_only completely, and only rely on corresponding banks reduce only. 

Another thought, I have not even thought about what it means if quote token is non usd for serum market hmm.

I was also thinking about deposit_to_vault_ratio, and I think it should not apply when bank has reduce only mode.

Reply:
as discussed,  deposit_to_vault_ratio is independent of reduce only mode